E2E tests with Playwright

castelblanque opened the pull request:
### Description of the change

E2E tests migrated to Playwright. Still some random failures, though.
One good reason to start using Playwright is its transparency when it comes to errors and reporting. That should allow us to pinpoint better the errors when they occur.

This PR also tries to address the following items regarding CI:

- Increased resources for the tests runner pod
- Increased machine used for e2e tests in CircleCI (from `medium` to `large`)
- Added a retry for every test. Playwright allows to re-run the entire test X times if it fails. Set to 1.
- Added reports of failing tests into generated artifacts of CircleCI: video of the whole process, last screenshot and a trace that can be viewed [here](https://trace.playwright.dev/).
![image](https://user-images.githubusercontent.com/67455978/151583863-739953a9-d49b-4333-9fd7-845add4ca2f9.png)

- Pre-loading Apache images into Kind (It was throwing 401 Unauthorized sometimes)
- Set up tests to be run with Chromium

Apart from that, the following non-repo related tasks have been done:

- Created new `kubeapps/integration-tests-pw` image for running Playwright tests
- Checked available node resources of main cluster in CI: it seems that still has resources available, I mean margin for holding more workload.

### Benefits

- Cleaner syntax and shorter tests
- Better reporting and logging of errors. E.g. videos are recorded, HTML report can be generated
- Automatic retries of tests
- We could easily expand testing to Firefox and Webkit too

### Possible drawbacks

Some random failures appear, still to be investigated with the new reporting data.

### Applicable issues

- Addresses #2898

absoludity commented on integration/tests/02-create-package-repository.spec.js:
```diff
@@ -0,0 +1,36 @@
+  // Clean up
+  const axInstance = await utils.getAxiosInstance(page);
+  const response = await axInstance.delete(
+    "/api/v1/clusters/default/namespaces/kubeapps/apprepositories/my-repo",
```
Hmm, unless you haven't merged main lately, this should not succeed, since the k8s api is no longer being proxied (by default)?

castelblanque replied:
This works due to the CI pipeline running with Operators and k8s APIs enabled.
We will have to do the delete operation using the UI. I added some clean up actions to keep the cluster freed from having many workloads due to limited resources.

absoludity replied:
> This works due to the CI pipeline running with Operators and k8s APIs enabled.

Yeah, I just meant that if you merge main, the `e2e-test.sh` script should no longer do that, but rather run all the main tests without the k8s APIs enabled, then install the operator support, upgrade kubeapps and run just the operator test.

> We will have to do the delete operation using the UI. I added some clean up actions to keep the cluster freed from having many workloads due to limited resources.

Yep, ok.

castelblanque replied:
Fixed that problem, now operators test is run correctly after all main tests.

absoludity commented on integration/tests/01-missing-permissions.spec.js:
```diff
@@ -0,0 +1,36 @@
+  await expect(errorLocator).toHaveCount(1);
+  await page.waitForTimeout(5000);
+
+  // For some reason, UI is showing different error messages randomly
```
Ah - I was wondering why you weren't testing for the presence of a particular error above. Keen to understand what is random here. What are some examples of what you see for the error message in the console below? Approx. what frequency etc?

castelblanque replied:
We are getting [both messages shown here](https://github.com/kubeapps/kubeapps/issues/4102).
Basically one is "unable to get secrets" and the other is "unable to create secrets". Most of the times one of them is thrown (don't remember which one from the top of my head), but every once in a while, the other is appearing instead. This is not related with testing, it is an issue that we should investigate.

absoludity reviewed (commented):
Great progress! Can the old files (test-cases) be removed? What is different between those and here?

castelblanque commented:
> Great progress! Can the old files (test-cases) be removed? 

Sure, will do!

> What is different between those and here?

Actions taken on old and new tests are the same so far. Just changed a bit the order and added a specific test for login with OIDC. The latter is added as first test, as I saw the login process failing many times with Puppeteer tests (e.g. stuck on DEX).

absoludity commented:
> @absoludity Tests seem to run smooth now (still taking a close look in CircleCI). [Last runs here](https://app.circleci.com/pipelines/github/kubeapps/kubeapps?branch=2898-e2e-test-engine-playwright&filter=all).

Excellent.

> 
> Not sure why, but navigating directly through URLs (e.g. `await page.goto(utils.getUrl("/#/c/default/ns/kubeapps/config/repos"));`) does not work very well. It works better to actually click on elements. Maybe is tests running to fast? maybe React app is unable to load correctly?

Hmm, not sure, but normally we'd want to click directly anyway right? I can look into why navigating directly has issues if we need to do so.

> 
> If you are going to take a look at permissions test, please notice that its new path is `/integration/tests/main/01-missing-permissions.spec.js`.

Do you mean that you're now happy to land this and do the release with the playwright tests? (Otherwise wouldn't I be looking at the missing perms in main where it has the old number). Happy to have this landed and do it there instead, if Antonio agrees too. Let me know. I'll start looking today either way.

absoludity commented on integration/tests/main/02-create-package-repository.spec.js:
```diff
@@ -0,0 +1,35 @@
+  await page.click('a:has-text("my-repo")');
+  await page.waitForSelector('css=.catalog-container .card-title >> text="gitlab-runner"');
+
+  // Clean up
+  const axInstance = await utils.getAxiosInstance(page);
+  const response = await axInstance.delete(
+    "/api/v1/clusters/default/namespaces/kubeapps/apprepositories/my-repo",
+  );
+  expect(response.status).toEqual(200);
```
Heh - at first I thought this shouldn't be possible as I misread and thought it was hitting the k8s API directly, but after checking, see this is a URI that's forwarded to the kubeops backend.

Still, that won't be available for ever either, why not use the UI to delete the repo as well? (Just like you're already doing in the private repo example below?)

castelblanque replied:
Agree, will move that to UI actions.

absoludity commented on integration/tests/main/03-create-private-package-repository.spec.js:
```diff
@@ -0,0 +1,130 @@
+const { test, expect } = require("@playwright/test");
```
Just noticed we're missing the SPDX headers with which the old tests were already updated.

castelblanque replied:
You are right, will add them.

absoludity commented on integration/tests/main/03-create-private-package-repository.spec.js:
```diff
@@ -0,0 +1,130 @@
+const utils = require("../utils/util-functions");
+
+test("Create a new private package repository successfully", async ({ page }) => {
+  test.setTimeout(120000);
```
I don't see this timeout in the old test? If it's necessary, just provide a comment explaining where the issue is so it can be investigated more easily later?

castelblanque replied:
This does not set a timeout to be effective at that point. It just sets the [timeout for this single test](https://playwright.dev/docs/test-timeouts#set-timeout-for-a-single-test). If the test takes more than 120 seconds, it will be considered as failed. There is a default that can be set on Playwright settings.

absoludity reviewed (approved):
Excellent, thanks Rafa. Comments in line, but nothing blocking.

castelblanque commented on integration/jest.config.js:
```diff
@@ -4,8 +4,8 @@
-  globalSetup: "jest-environment-puppeteer/setup",
-  globalTeardown: "jest-environment-puppeteer/teardown",
+/*   globalSetup: "jest-environment-puppeteer/setup",
+  globalTeardown: "jest-environment-puppeteer/teardown", */
```
Yep!

castelblanque commented on integration/Dockerfile:
```diff
@@ -5,5 +5,6 @@
 FROM bitnami/node:16.13.2
 RUN install_packages ca-certificates fonts-liberation gconf-service libappindicator1 libappindicator3-1 libasound2 libatk1.0-0 libatk-bridge2.0-0 libc6 libcairo2 libcups2 libdbus-1-3 libdrm2 libexpat1 libfontconfig1 libgbm1 libgbm-dev libgcc1 libgconf-2-4 libgdk-pixbuf2.0-0 libglib2.0-0 libgtk-3-0 libice6 libnspr4 libnss3 libpango-1.0-0 libpangocairo-1.0-0 libsm6 libstdc++6 libx11-6 libx11-xcb1 l ...
 ADD . /app/
+RUN rm -rf /app/tests
```
Yes, that is the purpose, otherwise image is shipped with tests. Added the comment in the commit msg as `Keep integration image free of test files` but will add it in the file if you want.

castelblanque commented on docs/user/manifests/kubeapps-local-dev-dex-values.yaml:
```diff
@@ -119,3 +119,10 @@ volumes:
 volumeMounts:
   - mountPath: /etc/dex/tls/https/server
     name: https-tls
+resources:
+  limits:
+    cpu: 500m
+    memory: 256Mi
+  requests:
+    cpu: 50m
+    memory: 64Mi
```
At least with Puppeteer, tests were failing many times while doing login with OIDC. Navigation would be lost at Dex side sometimes, so I decided to give it some resources. There were neither `limits` nor `requests` before.
Not sure if the change really had an effect, as many things were tried together (time constraints).
I'm totally fine with trying lower limits.

absoludity replied:
Fine with those limits, just think it'd be good to communicate to our future selves that these numbers aren't determined empirically. Maybe just something like:

```
# TODO(castelblanque): Added more resources because navigation would be lost at Dex sometimes. Not yet determined
# this had an effect nor whether the values are accurate.
```

to make it easier for someone else to make a decision when working here.

castelblanque commented on integration/manifests/executor.yaml:
```diff
@@ -21,8 +21,12 @@ spec:
           resources:
             requests:
-              memory: 100Mi
+              memory: 300Mi
+              cpu: 250m
+            limits:
+              memory: 600Mi
+              cpu: 500m
```
Setting resource requests/limitations is never an exact science. Given that now in CircleCI we have access to a (slightly) better machine plus the random problems we were facing, I decided to increase the requested resources.
Ideally, we would need to check empirically which resources are used on every pod throughout the CI process, and give them a bit more room for the future, etc.

castelblanque replied:
Integration pod is executed inside the kind cluster, so it might "eat" resources from the Kubeapps installation if limits are not set. Probably it is an edge case, but I think it is a good practice to set some boundaries to resources consumption.
BTW I have seen this (new?) [feature of CircleCI](https://app.circleci.com/pipelines/github/kubeapps/kubeapps/5945/workflows/f409f42c-d658-4345-844b-a6b22ecd927b/jobs/73076/resources) that shows the resources usage of Docker.

castelblanque commented on integration/package.json:
```diff
@@ -6,21 +6,16 @@
-    "start:window": "INTEGRATION_HEADLESS=false jest --runInBand",
-    "prettier": "prettier --write './**/*.{js, json}'"
+    "prettier": "prettier --write './**/*.{js, json}'",
+    "test": "playwright test --workers=1"
```
Yeah, in a later phase we can take a look at parallelism for tests 😃

castelblanque commented on integration/tests/main/02-create-package-repository.spec.js:
```diff
@@ -0,0 +1,35 @@
+  await k.doLogin("kubeapps-operator@example.com", "password", process.env.ADMIN_TOKEN);
+
+  // Go to repos page
+  await page.goto(utils.getUrl("/#/c/default/ns/kubeapps/config/repos"));
```
I have set a fixed `/` to start the login so that we keep things simple and clean.
Playing around with `page.goto` introduces randomness into the mix. For example, did try to set a test with a couple of `page.goto` with different urls: sometimes tests would run fine, and sometimes navigation would be reverted to the  homepage.

castelblanque commented on integration/tests/main/00-log-in.spec.js:
```diff
@@ -0,0 +1,13 @@
+const utils = require("../utils/util-functions");
+
+test.describe("Log in", () => {
+  test("Logs in successfully via OIDC", async ({ page }) => {
```
Good catch, it was the first test I did when there was not env var taken into consideration 😄

castelblanque commented:
> Hmm, not sure, but normally we'd want to click directly anyway right? I can look into why navigating directly has issues if we need to do so.

It depends for repetitive actions that are tested in previous tests it would be easier to navigate directly to it, for instance searching a text within a specific repo with a single `goto` url instead of the different UI actions to reach it.

> Do you mean that you're now happy to land this and do the release with the playwright tests? (Otherwise wouldn't I be looking at the missing perms in main where it has the old number). Happy to have this landed and do it there instead, if Antonio agrees too. Let me know. I'll start looking today either way.

I will add some comments and do some clean-up, and I would say it is fine for being used and hopefully improve our CI success rate.
Thanks for the review!

castelblanque commented on integration/tests/main/00-log-in.spec.js:
```diff
@@ -0,0 +1,13 @@
+test.describe("Log in", () => {
+  test("Logs in successfully via OIDC", async ({ page }) => {
+    const k = new KubeappsLogin(page);
+    await k.doLogin("kubeapps-user@example.com", "password", process.env.VIEW_TOKEN);
```
Agree. There is room for improvement in the testing side, I could not tackle everything. Goal for this was basically getting out of trouble with CI.
Is it ok if we create an issue with improvements for the future?

absoludity replied:
Yep, of course. Or even just a TODO comment here (so the next person who works on this can do a small drive-by improvement.

castelblanque commented:
> Happy to have this landed and do it there instead, if Antonio agrees too.

PR is ready to be merged. @antgamdia do you agree to proceed?

antgamdia commented on integration/tests/utils/util-functions.js:
```diff
@@ -0,0 +1,65 @@
+// Copyright 2021-2022 the Kubeapps contributors.
+// SPDX-License-Identifier: Apache-2.0
+
```
Are we using every function defined in this file? If not, let's remove them, no?

antgamdia commented on script/e2e-test.sh:
```diff
@@ -401,7 +396,7 @@ view_token="$(kubectl get -n kubeapps secret "$(kubectl get -n kubeapps servicea
 
 info "Running main Integration tests without k8s API access..."
-if ! kubectl exec -it "$pod" -- /bin/sh -c "INTEGRATION_RETRY_ATTEMPTS=3 INTEGRATION_ENTRYPOINT=http://kubeapps-ci.kubeapps USE_MULTICLUSTER_OIDC_ENV=${USE_MULTICLUSTER_OIDC_ENV} ADMIN_TOKEN=${admin_token} VIEW_TOKEN=${view_token} EDIT_TOKEN=${edit_token} yarn start ${ignoreFlag}"; then
+if ! kubectl exec -it "$pod" -- /bin/sh -c "CI_TIMEOUT=40 INTEGRATION_ENTRYPOINT=http://kubeapps-ci.kubeapps USE_MULTICLUSTER_OIDC_ENV=${USE_MULTICLUSTER_OIDC_ENV} ADMIN_TOKEN=${admin_token} VIEW_TOKEN=${view_token} EDIT_TOKEN=${edit_token} yarn test ${testsArgs}"; then
```
What this `CI_TIMEOUT` is doing?

castelblanque replied:
It is setting the global timeout for the execution. [See here.](https://github.com/kubeapps/kubeapps/blob/2898-e2e-test-engine-playwright/integration/playwright.config.js#L10)

antgamdia commented on .circleci/config.yml:
```diff
@@ -384,7 +384,12 @@ install_additional_cluster: &install_additional_cluster
+  run:
+    name: "Load needed images into Kind"
+    command:
+      ./script/load-kind-image.sh docker.io/bitnami/apache:2.4.48-debian-10-r74 kubeapps-ci kubeapps-ci-additional &&
+      ./script/load-kind-image.sh docker.io/bitnami/apache:2.4.48-debian-10-r75 kubeapps-ci kubeapps-ci-additional
```
Question: why is it required? I mean, shouldn't the cluster pull the images if necessary?

castelblanque replied:
It turns out that there were some runs where tests failed due to those images not being present and failing to be pulled upon pod creation. FYI, at some runs  `2.4.48-debian-10-r74` would be successfully pulled but `2.4.48-debian-10-r75` would fail with a `401`!. Since images are pre-pushed I haven't seen an error regarding this.

antgamdia reviewed (approved):
+1, just minor comments and questions, but happy to land and continue iterating over it.

castelblanque commented on integration/tests/utils/kubeapps-login.js:
```diff
@@ -0,0 +1,81 @@
+    await formLocator.type("input[name=token]", token);
+    await this.page.click("#login-submit-button");
+
+    await this.page.waitForLoadState("networkidle");
+    await this.page.waitForLoadState("domcontentloaded");
```
Login was being a painpoint in previous tests, many times due to page not being in a consistent state after login was done.
From all the runs that I have executed, having both `networkidle` and `domcontentloaded` were the more stable ones, mostly not failing. And IMHO it makes sense to have both, otherwise we might run in having the dom loaded but still ongoing network calls or maybe no network connections (I believe it checks the last 500ms) but the dom not yet loaded (for any reason). On top of that, it doesn't cause any harm and as said, it brings stability.

castelblanque commented on integration/tests/main/07-upgrade.spec.js:
```diff
@@ -0,0 +1,72 @@
+  await k.doLogin("kubeapps-operator@example.com", "password", process.env.EDIT_TOKEN);
+
+  // Go to catalog
+  //await page.goto(utils.getUrl("/#/c/default/ns/default/catalog?Search=apache&Repository=bitnami"));
```
Agree, as mentioned in other comments, having multiple `page.goto` in the same test (login does one already) produces errors sometimes.

castelblanque commented on integration/playwright.config.js:
```diff
@@ -0,0 +1,30 @@
+// Copyright 2021-2022 the Kubeapps contributors.
```
Just copy-pasted the existing headers from the Puppeteer tests. Will remove `2021`.

castelblanque commented on integration/package.json:
```diff
@@ -6,21 +6,16 @@
-    "start": "jest --runInBand --ci --colors",
-    "start:window": "INTEGRATION_HEADLESS=false jest --runInBand",
-    "prettier": "prettier --write './**/*.{js, json}'"
+    "prettier": "prettier --write './**/*.{js, json}'",
```
Haven't changed that actually. It is behaving the same way it did with Puppeteer tests. It wasn't the goal for this issue, we can create a new issue to fix/fine tune if needed.

castelblanque commented on integration/package.json:
```diff
@@ -6,21 +6,16 @@
+    "test": "playwright test --workers=1"
   },
   "dependencies": {
-    "@jest/test-sequencer": "^27.4.6",
```
We are having a single worker, no parallelism for running tests. In this case, by default Playwright states `if you disable parallelism, you can control test order by either naming your files in alphabetical order or using a "test list" file.`. We have test files starting with numbers, so we should be good.
Only caveat is that tests are grouped now in folders (e.g. `main` and `multicluster`), and folders are run in order. If we needed a test from folder A to be dependent on test from folder B...that would require some adjustments 😄

castelblanque commented on integration/Dockerfile:
```diff
@@ -5,5 +5,8 @@
 FROM bitnami/node:16.13.2
 RUN install_packages ca-certificates fonts-liberation gconf-service libappindicator1 libappindicator3-1 libasound2 libatk1.0-0 libatk-bridge2.0-0 libc6 libcairo2 libcups2 libdbus-1-3 libdrm2 libexpat1 libfontconfig1 libgbm1 libgbm-dev libgcc1 libgconf-2-4 libgdk-pixbuf2.0-0 libglib2.0-0 libgtk-3-0 libice6 libnspr4 libnss3 libpango-1.0-0 libpangocairo-1.0-0 libsm6 libstdc++6 libx11-6 libx11-xcb1 l ...
```
Definitely! As this task will take some time, will do it in a different PR if you don't mind.

castelblanque commented on .circleci/config.yml:
```diff
@@ -687,6 +693,7 @@ jobs:
             ./script/create_release.sh ${CIRCLE_TAG} << pipeline.parameters.KUBEAPPS_REPO >>
   local_e2e_tests:
     machine: true
+    resource_class: large
```
Added the URL to CircleCI docs.
Regarding #4170, if you think that it will help with "spending" better the minutes in CircleCI, I would say go for it. Probably we will need them 😄

antgamdia commented:
:tada:

castelblanque commented:
😅
